pkg/pool-weighted/contracts/smart/InvestmentPool.sol

```diff
@@ -0,0 +1,271 @@
+    uint256 private constant _END_WEIGHT_OFFSET = 64;
+    uint256 private constant _DECIMAL_DIFF_OFFSET = 96;
+
+    // Time travel comment
+    // [ 192 bits | 32 bits  |  32 bits   |
+    // [  unused  | end time | start time |
+    // |MSB                            LSB|
+    //bytes32 private _gradualUpdateTimestamps;
+
+    //uint256 private constant _START_TIME_OFFSET = 0;
+    //uint256 private constant _END_TIME_OFFSET = 32;
+
+    constructor(
+        IVault vault,
+        string memory name,
+        string memory symbol,
+        IERC20[] memory tokens,
+        uint256[] memory normalizedWeights,
+        uint256 swapFeePercentage,
+        uint256 pauseWindowDuration,
+        uint256 bufferPeriodDuration,
+        address owner
+    )
+        BaseWeightedPool(
+            vault,
+            name,
+            symbol,
+            tokens,
+            new address[](tokens.length), // Pass the zero address: Investment Pools can't have asset managers
+            swapFeePercentage,
+            pauseWindowDuration,
+            bufferPeriodDuration,
+            owner
+        )
+    {
+        uint256 numTokens = tokens.length;
+        InputHelpers.ensureInputLengthMatch(numTokens, normalizedWeights.length);
+
+        _totalTokens = numTokens;
+
+        // I'm time-traveling a bit here - storing the weights in a form where they can be changed
+        uint256 currentTime = block.timestamp;
+        _startGradualWeightChange(currentTime, currentTime, normalizedWeights, normalizedWeights, tokens);
+    }
+
+    function _getMaxTokens() internal pure virtual override returns (uint256) {
+        return _MAX_WEIGHTED_TOKENS;
+    }
+
+    function _getTotalTokens() internal view virtual override returns (uint256) {
+        return _totalTokens;
+    }
+
+    function _scalingFactor(IERC20 token) internal view virtual override returns (uint256) {
+        bytes32 tokenData = _poolState[token];
+
+        // A valid token can't be zero (must have non-zero weights)
+        if (tokenData == 0) {
+            _revert(Errors.INVALID_TOKEN);
+        }
+
+        return _computeScalingFactor(tokenData);
+    }
+
+    function _scalingFactors() internal view virtual override returns (uint256[] memory scalingFactors) {
+        (IERC20[] memory tokens, , ) = getVault().getPoolTokens(getPoolId());
+        uint256 numTokens = tokens.length;
+
+        scalingFactors = new uint256[](numTokens);
+
+        for (uint256 i = 0; i < numTokens; i++) {
+            scalingFactors[i] = _computeScalingFactor(_poolState[tokens[i]]);
+        }
+    }
+
+    function _getNormalizedWeight(IERC20 token) internal view override returns (uint256) {
+        bytes32 tokenData = _poolState[token];
+
+        // A valid token can't be zero (must have non-zero weights)
+        if (tokenData == 0) {
+            _revert(Errors.INVALID_TOKEN);
+        }
+
+        uint256 startWeight = tokenData.decodeUint64(_START_WEIGHT_OFFSET).uncompress64();
```

Comment:
Same here, I think this bit should be broken out into a function (similar to how it's done by index in the LBP, it can be done by passing tokenData and pctProgress here):

```
uint256 startWeight = tokenData.decodeUint64(_START_WEIGHT_OFFSET).uncompress64();
uint256 endWeight = tokenData.decodeUint32(_END_WEIGHT_OFFSET).uncompress32();

return _interpolateWeight(startWeight, endWeight, pctProgress);
```

I just don't like having so many lines duplicated across functions because it makes me want to double-check that they are actually the same, which is a waste of time/energy. If there's a good reason not to do this please let me know.

Reply:
Since this is only called before interpolateWeight, I'll just move it inside there.